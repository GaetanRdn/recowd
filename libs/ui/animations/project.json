{
  "projectType": "library",
  "sourceRoot": "libs/ui/animations/src",
  "prefix": "rc",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:ng-packagr-lite",
      "outputs": [
        "dist/libs/ui/animations"
      ],
      "options": {
        "project": "libs/ui/animations/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/ui/animations/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/ui/animations/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/ui/animations/**/*.ts",
          "libs/ui/animations/**/*.html"
        ]
      }
    }
  },
  "tags": [
    "ui"
  ]
}
