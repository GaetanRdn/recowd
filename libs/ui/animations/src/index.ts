export * from './lib/rotate';
export * from './lib/zoom';
