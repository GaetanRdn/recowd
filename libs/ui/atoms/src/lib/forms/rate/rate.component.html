<rc-material-icon
  *ngFor="let nb of range"
  (click)="setValue(nb)"
  [class.rc-selected]="value >= nb"
  [@zoom]="value >= nb ? 'zoomed' : 'default'"
>
  {{ nb <= value ? 'star_rate' : 'star_border' }}
</rc-material-icon>
