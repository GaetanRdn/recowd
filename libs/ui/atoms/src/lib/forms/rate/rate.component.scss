@use '../../assets/utils';

@mixin rc-rate($config) {
  $colors: map-get($config, colors);

  rc-rate {
    &:not(.rc-disabled) {
      cursor: pointer;
    }

    &.rc-disabled {
      rc-material-icon {
        color: utils.map-get-deep($colors, disabled, color);
      }
    }
  }
}
