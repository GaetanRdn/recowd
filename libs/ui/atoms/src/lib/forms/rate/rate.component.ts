import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  Input,
  NgModule,
  OnChanges,
  Output,
} from '@angular/core';
import { MaterialIconModule } from '../../material-icon/material-icon.component';
import { CommonModule } from '@angular/common';
import { BooleanInput, CoerceBoolean, NgChanges } from '@recowd/utility-types';
import { zoomAnimation } from '@recowd/ui-animations';

@Component({
  selector: 'rc-rate',
  host: {
    '[class.rc-disabled]': 'disabled',
  },
  templateUrl: './rate.component.html',
  styleUrls: ['./rate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  animations: [zoomAnimation],
})
export class RateComponent implements OnChanges {
  static ngAcceptInputType_disabled: BooleanInput;

  @Input() public nbStars = 1;

  @Input() public value = 0;

  @Input() @CoerceBoolean() public disabled = false;

  @Output() public readonly valueChange: EventEmitter<number> =
    new EventEmitter<number>();

  get range(): number[] {
    return Array.from(Array(this.nbStars).keys()).map((nb: number) => nb + 1);
  }

  public ngOnChanges(changes: NgChanges<RateComponent>): void {
    if (changes.nbStars && changes.nbStars.currentValue === 0) {
      throw new Error('this nbStars should be > 0');
    }
    if (
      changes.value &&
      (changes.value.currentValue < 0 ||
        changes.value.currentValue > this.nbStars)
    ) {
      throw new Error(`this value is out of range [0, ${this.nbStars}]`);
    }
  }

  public setValue(nb: number): void {
    if (!this.disabled) {
      if (this.value !== nb) {
        this.value = nb;
      } else {
        this.value = 0;
      }
      this.valueChange.emit(this.value);
    }
  }
}

@NgModule({
  declarations: [RateComponent],
  exports: [RateComponent],
  imports: [MaterialIconModule, CommonModule],
})
export class RateModule {}
