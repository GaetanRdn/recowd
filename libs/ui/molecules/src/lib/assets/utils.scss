@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
