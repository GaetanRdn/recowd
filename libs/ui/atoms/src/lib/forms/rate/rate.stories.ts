import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { BADGE } from '@geometricpanda/storybook-addon-badges';
import { RateComponent, RateModule } from './rate.component';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';

export default {
  title: 'atoms/forms/rate',
  component: RateComponent,
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'rate.component',
  },
  decorators: [
    moduleMetadata({
      imports: [RateModule, BrowserAnimationsModule],
    }),
  ],
} as Meta<RateComponent>;

const template: Story<RateComponent> = (args: RateComponent) => ({
  props: {
    nbStars: args['nbStars'],
    value: args['value'],
    disabled: args['disabled'],
  },
});

export const Default = template.bind({});
Default.args = {
  nbStars: 1,
  value: 0,
  disabled: false,
};
