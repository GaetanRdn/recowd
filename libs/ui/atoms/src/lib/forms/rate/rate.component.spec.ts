import { TestBed, waitForAsync } from '@angular/core/testing';

import { RateComponent } from './rate.component';
import { TemplateLookup } from '@recowd/test/utils';
import { Component, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';

describe('RateComponent', () => {
  let templateLookup: TemplateLookup<HostComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [HostComponent],
      imports: [NoopAnimationsModule, RateComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();
  }));

  beforeEach(() => (templateLookup = new TemplateLookup(HostComponent)));

  it('should create', () => {
    templateLookup.detectChanges();
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  test.each([3, 5])('should create with %s stars', (nbStars: number) => {
    templateLookup.hostComponent.nbStars = nbStars;
    templateLookup.detectChanges();
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  test('should throw an error with 0 star', () => {
    templateLookup.hostComponent.nbStars = 0;
    expect(() => templateLookup.detectChanges()).toThrowError(
      'this nbStars should be > 0'
    );
  });

  test.each([3, 5])('should create with %s filled stars', (value: number) => {
    templateLookup.hostComponent.nbStars = 5;
    templateLookup.hostComponent.value = value;
    templateLookup.detectChanges();
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  test.each`
    filled | nbStars
    ${-1}  | ${3}
    ${6}   | ${5}
  `(
    'should throw error with $filled filled stars and 5 nb stars',
    ({ filled, nbStars }) => {
      templateLookup.hostComponent.nbStars = nbStars;
      templateLookup.hostComponent.value = filled;
      expect(() => templateLookup.detectChanges()).toThrowError(
        `this value is out of range [0, ${nbStars}]`
      );
    }
  );

  test.each([1, 4])('should select %s stars', (value: number) => {
    // GIVEN
    templateLookup.hostComponent.nbStars = 5;
    templateLookup.detectChanges();

    // WHEN
    templateLookup.queryAll('rc-material-icon')[value - 1].click();
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.getComponent(RateComponent).value).toEqual(value);
    expect(templateLookup.hostComponent.value).toEqual(value);
  });

  test.each([1, 4])('should deselect %s stars', (value: number) => {
    // GIVEN
    templateLookup.hostComponent.nbStars = 5;
    templateLookup.hostComponent.value = value;
    templateLookup.detectChanges();

    // WHEN
    templateLookup.queryAll('rc-material-icon')[value - 1].click();
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.getComponent(RateComponent).value).toEqual(0);
    expect(templateLookup.hostComponent.value).toEqual(0);
  });

  it('should create disabled', () => {
    // GIVEN
    templateLookup.hostComponent.disabled = true;

    // WHEN
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
  });

  it('should not select when disabled', () => {
    // GIVEN
    templateLookup.hostComponent.disabled = true;
    templateLookup.detectChanges();

    // WHEN
    templateLookup.queryAll('rc-material-icon')[0].click();
    templateLookup.detectChanges();

    // THEN
    expect(templateLookup.firstChildElement).toMatchSnapshot();
    expect(templateLookup.getComponent(RateComponent).value).toEqual(0);
    expect(templateLookup.hostComponent.value).toEqual(0);
  });
});

@Component({
  template: ` <rc-rate
    [nbStars]="nbStars"
    [(value)]="value"
    [disabled]="disabled"
  ></rc-rate>`,
})
class HostComponent {
  public nbStars?: RateComponent['nbStars'];
  public value: RateComponent['value'] = 0;
  public disabled: RateComponent['disabled'] = false;
}
