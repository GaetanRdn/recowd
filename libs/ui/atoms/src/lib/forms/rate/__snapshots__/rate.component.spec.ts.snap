// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateComponent should create 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should create disabled 1`] = `
<rc-rate
  class="rc-disabled"
>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should create with 3 filled stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should create with 3 stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should create with 5 filled stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="rc-selected ng-trigger ng-trigger-zoom"
    style=""
  >
     star_rate

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should create with 5 stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should deselect 1 stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom ng-animating"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should deselect 4 stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom ng-animating"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom ng-animating"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom ng-animating"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom ng-animating"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should not select when disabled 1`] = `
<rc-rate
  class="rc-disabled"
>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should select 1 stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom rc-selected ng-animating"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;

exports[`RateComponent should select 4 stars 1`] = `
<rc-rate>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom rc-selected ng-animating"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom rc-selected ng-animating"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom rc-selected ng-animating"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom rc-selected ng-animating"
    style=""
  >
     star_rate

  </rc-material-icon>
  <rc-material-icon
    class="ng-trigger ng-trigger-zoom"
    style=""
  >
     star_border

  </rc-material-icon>
  
</rc-rate>
`;
